Makefile

```diff
@@ -1,82 +1,93 @@
-CFLAGS=-g 
+CC=g++
+CFLAGS=-g
+SERVER=server
+EMULATOR=msp430-emu
 
-all: MSP430 SERVER
+.PHONY: all test
 
-MSP430 : main.o utilities.o emu_server.o registers.o memspace.o debugger.o disassembler.o \
-	register_display.o decoder.o flag_handler.o formatI.o formatII.o formatIII.o \
-	usci.o port1.o packet_queue.o bcm.o timer_a.o
+all: ${EMULATOR} ${SERVER}
 
-	g++ ${CFLAGS} -o MSP430 main.o emu_server.o utilities.o registers.o memspace.o debugger.o disassembler.o \
-	register_display.o decoder.o flag_handler.o formatI.o formatII.o formatIII.o usci.o port1.o bcm.o timer_a.o packet_queue.o -lreadline -lwebsockets -lpthread -lrt -lssl -lcrypto;
+# Main emulator program
+
+${EMULATOR} : main.o emu_server.o utilities.o registers.o memspace.o debugger.o disassembler.o \
+	register_display.o decoder.o flag_handler.o formatI.o formatII.o formatIII.o usci.o port1.o bcm.o \
+	timer_a.o packet_queue.o io.o
+	${CC} ${CFLAGS} -o $@ $^ -lreadline -lwebsockets -lpthread -lrt -lssl -lcrypto;
 
 main.o : main.cpp
-	g++ ${CFLAGS} -c main.cpp
+	${CC} ${CFLAGS} -c $<
 
 utilities.o : devices/utilities.c
-	g++ ${CFLAGS} -c devices/utilities.c
+	${CC} ${CFLAGS} -c $<
 
 registers.o : devices/cpu/registers.c
-	g++ ${CFLAGS} -c devices/cpu/registers.c
+	${CC} ${CFLAGS} -c $<
 
 memspace.o : devices/memory/memspace.c
-	g++ ${CFLAGS} -c devices/memory/memspace.c
+	${CC} ${CFLAGS} -c $<
 
 debugger.o : debugger/debugger.c
-	g++ ${CFLAGS} -c debugger/debugger.c
+	${CC} ${CFLAGS} -c $<
 
 disassembler.o : debugger/disassembler.c
-	g++ ${CFLAGS} -c debugger/disassembler.c
+	${CC} ${CFLAGS} -c $<
 
 register_display.o : debugger/register_display.c
-	g++ ${CFLAGS} -c debugger/register_display.c
+	${CC} ${CFLAGS} -c $<
 
 decoder.o : devices/cpu/decoder.c
-	g++ ${CFLAGS} -c devices/cpu/decoder.c
+	${CC} ${CFLAGS} -c $<
 
 flag_handler.o : devices/cpu/flag_handler.c
-	g++ ${CFLAGS} -c devices/cpu/flag_handler.c
+	${CC} ${CFLAGS} -c $<
 
 formatI.o : devices/cpu/formatI.c
-	g++ ${CFLAGS} -c devices/cpu/formatI.c
+	${CC} ${CFLAGS} -c $<
 
 formatII.o : devices/cpu/formatII.c
-	g++ ${CFLAGS} -c devices/cpu/formatII.c
+	${CC} ${CFLAGS} -c $<
 
 formatIII.o : devices/cpu/formatIII.c
-	g++ ${CFLAGS} -c devices/cpu/formatIII.c
+	${CC} ${CFLAGS} -c $<
 
 bcm.o : devices/peripherals/bcm.c
-	g++ ${CFLAGS} -c devices/peripherals/bcm.c
+	${CC} ${CFLAGS} -c $<
 
 timer_a.o : devices/peripherals/timer_a.c
-	g++ ${CFLAGS} -c devices/peripherals/timer_a.c
+	${CC} ${CFLAGS} -c $<
 
 usci.o : devices/peripherals/usci.c
-	g++ ${CFLAGS} -c devices/peripherals/usci.c
+	${CC} ${CFLAGS} -c $<
```

Comment:
a lot of repeated rules, why not ` ‘%.o : %.c’ ` ?

Reply:
So that the overall structure is not that different from the original makefile...